src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexFacetedReadOperation.cs

```diff
@@ -0,0 +1,428 @@
+            for (int docId = 0; docId < read; docId++)
+            {
+                var entryReader = _indexSearcher.GetReaderFor(ids[docId]);
+                foreach (var result in results)
+                {
+                    token.ThrowIfCancellationRequested();
+
+                    using var facetTiming = queryTimings?.For($"{nameof(QueryTimingsScope.Names.AggregateBy)}/{result.Key}");
+
+                    if (result.Value.Ranges == null || result.Value.Ranges.Count == 0)
+                    {
+                        facetsByName ??= new Dictionary<string, Dictionary<string, FacetValues>>();
+
+                        HandleFacetsPerDocument(ref entryReader, result, facetsByName, facetQuery.Legacy, facetTiming, analyzersScope, token);
+                        continue;
+                    }
+
+                    // Cache facetByRange because we will fulfill data in batches instead of whole collection
+                    facetsByRange ??= new();
+                    if (facetsByRange.TryGetValue(result.Key, out var facetValues) == false)
+                    {
+                        facetValues = new();
+                        facetsByRange.Add(result.Key, facetValues);
+                    }
+
+                    HandleRangeFacetsPerDocument(ref entryReader, result, facetQuery.Legacy, facetTiming, facetValues, token);
+                }
+            }
+
+            token.ThrowIfCancellationRequested();
+        }
+
+        UpdateRangeResults();
+        
+        UpdateFacetResults(results, query, facetsByName);
+
+        CompleteFacetCalculationsStage(results);
+        CoraxIndexReadOperation.QueryPool.Return(ids);
+        return results.Values
+            .Select(x => x.Result)
+            .ToList();
+
+        void UpdateRangeResults()
+        {
+            foreach (var result in results)
+            {
+                if (facetsByRange != null)
+                    foreach (var kvp in facetsByRange)
+                    {
+                        foreach (var inner in kvp.Value)
+                        {
+                            if (inner.Value.Any == false)
+                                continue;
+
+                            result.Value.Result.Values.AddRange(inner.Value.GetAll());
+                        }
+                    }
+            }
+        }
+    }
+
+    private void HandleRangeFacetsPerDocument(ref IndexEntryReader indexEntry,
+        KeyValuePair<string, FacetedQueryParser.FacetResult> result,
+        bool legacy,
+        QueryTimingsScope queryTimings,
+        Dictionary<string, FacetValues> facetValues,
+        CancellationToken token)
+    {
+        var needToApplyAggregation = result.Value.Aggregations.Count > 0;
+        var ranges = result.Value.Ranges;
+        
+        // Create map in first batch
+        if (facetValues.Count == 0)
+        {
+            CreateFacetMapping();
+        }
+
+
+        foreach (var parsedRange in CollectionsMarshal.AsSpan(ranges))
```

Comment:
Why `AsSpan` ?

Reply:
removed. `AsSpan` could be faster than the normal `foreach` on a list.

Reply:
Why? A `foreach` on a list is reduced to a struct and no additional cost, no?

Reply:
https://www.meziantou.net/fastest-way-to-enumerate-a-list-t.htm